File: scripts/bulk_upload_customers.py
#!/usr/bin/env python3
"""
Satya Cable & Broadband Network - Firestore Bulk Upload Script (Python)

Features:
1. Bulk Operations: Uses Firestore Python SDK `bulk_writer()` to execute batched
   writes efficiently (up to 500 writes per batch automatically queued and flushed).
   
2. Service Types & Provider Hierarchy:
   - Service Types: 'cable', 'internet', 'internet+iptv', 'apfiber'
   - Providers per Service Type:
       • cable: 'act' (ACT Cable Network - Current accounts), 'sitv' (SITV Cable)
       • internet: 'vbc' (VBC Broadband), 'bsnl' (BSNL Fiber)
       • internet+iptv: 'fiber_iptv' (Fiber + IPTV Combo)
       • apfiber: 'apsfl' (AP FiberNet)

3. Collection Separation & Relations:
   - `customers`: Customer personal details. Contains `accountIds` array & `activeServices` array.
   - `customer_accounts`: Connection details. Contains `customerId` reference pointing back to `customers`.
   - `services`: Mirror collection of `customer_accounts` for compatibility.
   - `service_providers` & `packages`: Registry for providers & package details.

Usage:
   python3 scripts/bulk_upload_customers.py
"""

import os
import re
import json
from datetime import datetime, timezone
from google.cloud import firestore
from google.oauth2 import service_account

# -----------------------------------------------------------------------------
# Config & Credentials Setup
# -----------------------------------------------------------------------------
PROJECT_ID = os.environ.get("FIREBASE_PROJECT_ID", "cableapp-642c4")
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
ROOT_DIR = os.path.dirname(SCRIPT_DIR)
SERVICE_ACCOUNT_FILE = os.path.join(ROOT_DIR, "cableapp-642c4-firebase-adminsdk-fbsvc-ee5e58c939.json")

def get_now_iso():
    return datetime.now(timezone.utc).isoformat()

def get_db_client():
    sa_env = os.environ.get("GOOGLE_APPLICATION_CREDENTIALS")
    if sa_env and os.path.exists(sa_env):
        print(f"🔑 Authenticating using GOOGLE_APPLICATION_CREDENTIALS: {sa_env}")
        creds = service_account.Credentials.from_service_account_file(sa_env)
        return firestore.Client(project=PROJECT_ID, credentials=creds)
    elif os.path.exists(SERVICE_ACCOUNT_FILE):
        print(f"🔑 Authenticating using service account file: {SERVICE_ACCOUNT_FILE}")
        creds = service_account.Credentials.from_service_account_file(SERVICE_ACCOUNT_FILE)
        return firestore.Client(project=PROJECT_ID, credentials=creds)
    else:
        print(f"ℹ️ Initializing Firestore Client with default credentials for project: {PROJECT_ID}")
        return firestore.Client(project=PROJECT_ID)

# -----------------------------------------------------------------------------
# Service Providers Registry
# -----------------------------------------------------------------------------
SERVICE_PROVIDERS = [
    {
        "id": "cable_act",
        "serviceType": "cable",
        "serviceTypeName": "Cable TV",
        "provider": "act",
        "providerName": "ACT Cable Network",
        "isDefault": True,
        "packageIds": [],
        "updatedAt": get_now_iso()
    },
    {
        "id": "cable_sitv",
        "serviceType": "cable",
        "serviceTypeName": "Cable TV",
        "provider": "sitv",
        "providerName": "SITV Digital Cable",
        "isDefault": False,
        "packageIds": [],
        "updatedAt": get_now_iso()
    },
    {
        "id": "internet_vbc",
        "serviceType": "internet",
        "serviceTypeName": "High-Speed Broadband",
        "provider": "vbc",
        "providerName": "VBC Broadband Network",
        "isDefault": False,
        "packageIds": [],
        "updatedAt": get_now_iso()
    },
    {
        "id": "internet_bsnl",
        "serviceType": "internet",
        "serviceTypeName": "High-Speed Broadband",
        "provider": "bsnl",
        "providerName": "BSNL Bharat Fiber",
        "isDefault": False,
        "packageIds": [],
        "updatedAt": get_now_iso()
    },
    {
        "id": "internet_iptv_combo",
        "serviceType": "internet+iptv",
        "serviceTypeName": "Internet + IPTV Combo",
        "provider": "fiber_iptv",
        "providerName": "Fiber + IPTV Combo Digital",
        "isDefault": False,
        "packageIds": [],
        "updatedAt": get_now_iso()
    },
    {
        "id": "apfiber_apsfl",
        "serviceType": "apfiber",
        "serviceTypeName": "AP FiberNet (APSFL)",
        "provider": "apsfl",
        "providerName": "AP FiberNet Corporation",
        "isDefault": False,
        "packageIds": [],
        "updatedAt": get_now_iso()
    }
]

def get_pkg_slug(pkg_name):
    if not pkg_name:
        return "base_pack"
    slug = re.sub(r'[^a-z0-9]+', '_', pkg_name.lower()).strip('_')
    return slug or "base_pack"

def parse_report_html(file_path):
    with open(file_path, 'r', encoding='utf-8', errors='ignore') as f:
        html_content = f.read()

    tr_matches = re.findall(r'<tr[\s\S]*?</tr>', html_content, re.IGNORECASE)
    raw_accounts = []

    for tr in tr_matches:
        td_matches = re.findall(r'<td[\s\S]*?</td>', tr, re.IGNORECASE)
        if not td_matches or len(td_matches) < 10:
            continue

        clean_cells = []
        for cell in td_matches:
            text = re.sub(r'<[^>]+>', '', cell).strip()
            if text.startswith("'"):
                text = text[1:]
            clean_cells.append(text)

        acct_no = clean_cells[1] if len(clean_cells) > 1 else ""
        if not acct_no or acct_no.lower() == "account no":
            continue

        raw_accounts.append({
            "accountNumber": acct_no,
            "lcoCustomerId": clean_cells[2] if len(clean_cells) > 2 else "",
            "stbSerialNumber": clean_cells[3] if len(clean_cells) > 3 else "",
            "vcNumber": clean_cells[4] if len(clean_cells) > 4 else "",
            "name": clean_cells[5] if len(clean_cells) > 5 else "Unknown Customer",
            "address": clean_cells[6] if len(clean_cells) > 6 else "N/A",
            "createdDate": clean_cells[7] if len(clean_cells) > 7 else get_now_iso(),
            "phone": clean_cells[8] if len(clean_cells) > 8 else "919999999999",
            "packageName": clean_cells[9] if len(clean_cells) > 9 else "Base Pack",
            "status": "active" if "act" in (clean_cells[10].lower() if len(clean_cells) > 10 else "") else "inactive",
            "msoShareDue": float(clean_cells[11].replace(',', '')) if len(clean_cells) > 11 and clean_cells[11] else 0.0
        })

    return raw_accounts

def load_data():
    data_dir = os.path.join(SCRIPT_DIR, "data")
    cust_json = os.path.join(data_dir, "customers.json")
    svc_json = os.path.join(data_dir, "services.json")
    xls_path = os.path.join(SCRIPT_DIR, "Customer_account_status_report_2026-09-08 14_09_35.xls")

    if os.path.exists(cust_json) and os.path.exists(svc_json):
        print(f"📦 Reading dataset from JSON files in {data_dir}...")
        with open(cust_json, 'r', encoding='utf-8') as f:
            customers_raw = json.load(f)
        with open(svc_json, 'r', encoding='utf-8') as f:
            services_raw = json.load(f)

        packages_map = {}
        cust_accounts_map = {}
        cust_services_map = {}

        accounts_list = []
        for s in services_raw:
            pkg_name = s.get("packages", ["Base Pack"])[0] if s.get("packages") else "Base Pack"
            pkg_slug = s.get("packageIds", [get_pkg_slug(pkg_name)])[0] if s.get("packageIds") else get_pkg_slug(pkg_name)

            if pkg_slug not in packages_map:
                packages_map[pkg_slug] = {
                    "id": pkg_slug,
                    "packageName": pkg_name,
                    "serviceType": s.get("serviceCategory", "cable"),
                    "provider": s.get("provider", "act"),
                    "providerId": s.get("providerId", "cable_act"),
                    "providerName": s.get("providerName", "ACT Cable Network"),
                    "price": 0.0,
                    "isActive": True,
                    "createdAt": get_now_iso()
                }

            c_id = s.get("customerId")
            if c_id not in cust_accounts_map:
                cust_accounts_map[c_id] = []
                cust_services_map[c_id] = set()
            
            acct_id = s.get("accountNumber") or s.get("id")
            cust_accounts_map[c_id].append(acct_id)
            cust_services_map[c_id].add(s.get("serviceCategory", "cable"))

            accounts_list.append({
                "id": acct_id,
                "accountNumber": acct_id,
                "customerId": c_id,
                "lcoCustomerId": s.get("lcoCustomerId", ""),
                "stbSerialNumber": s.get("stbSerialNumber", ""),
                "vcNumber": s.get("vcNumber", ""),
                "serviceType": s.get("serviceCategory", "cable"),
                "serviceTypeName": s.get("serviceCategoryName", "Cable TV"),
                "provider": s.get("provider", "act"),
                "providerName": s.get("providerName", "ACT Cable Network"),
                "providerId": s.get("providerId", "cable_act"),
                "packageIds": [pkg_slug],
                "packages": [pkg_name],
                "status": s.get("status", "active"),
                "msoShareDue": s.get("msoShareDue", 0.0),
                "createdAt": s.get("createdAt", get_now_iso()),
                "updatedAt": get_now_iso()
            })

        customers_list = []
        for c in customers_raw:
            c_id = c["id"]
            customers_list.append({
                "id": c_id,
                "customerCode": c.get("customerCode", f"SAT-{c_id}"),
                "name": c.get("name", "Unknown Customer"),
                "phone": c.get("phone", "919999999999"),
                "address": c.get("address", "N/A"),
                "accountIds": cust_accounts_map.get(c_id, []),
                "activeServices": list(cust_services_map.get(c_id, {"cable"})),
                "createdAt": c.get("createdAt", get_now_iso()),
                "updatedAt": get_now_iso()
            })

        return customers_list, accounts_list, list(packages_map.values())
    elif os.path.exists(xls_path):
        print(f"📦 Parsing customer report XLS/HTML: {xls_path}...")
        raw_accounts = parse_report_html(xls_path)
        print(f"   Found {len(raw_accounts)} raw account records.")

        dummy_phones = {'919111111111', '919490712587', '', '919999999999'}
        dedup_map = {}
        cust_counter = 10001
        cust_dict = {}
        accounts_list = []
        packages_map = {}

        for item in raw_accounts:
            phone_clean = re.sub(r'[^0-9]', '', item["phone"])
            norm_phone = "" if phone_clean in dummy_phones else phone_clean
            norm_name = item["name"].lower().strip()
            norm_addr = item["address"].lower().strip()

            dedup_key = f"phone:{norm_phone}" if norm_phone else f"name_addr:{norm_name}_{norm_addr}"
            if dedup_key not in dedup_map:
                c_id = f"CUST-{cust_counter}"
                cust_counter += 1
                dedup_map[dedup_key] = c_id
                cust_dict[c_id] = {
                    "id": c_id,
                    "customerCode": f"SAT-{c_id}",
                    "name": item["name"],
                    "phone": item["phone"],
                    "address": item["address"],
                    "accountIds": [],
                    "activeServices": set(),
                    "createdAt": item["createdDate"],
                    "updatedAt": get_now_iso()
                }
            else:
                c_id = dedup_map[dedup_key]

            cust_dict[c_id]["accountIds"].append(item["accountNumber"])
            cust_dict[c_id]["activeServices"].add("cable")

            pkg_name = item["packageName"]
            pkg_slug = get_pkg_slug(pkg_name)
            if pkg_slug not in packages_map:
                packages_map[pkg_slug] = {
                    "id": pkg_slug,
                    "packageName": pkg_name,
                    "serviceType": "cable",
                    "provider": "act",
                    "providerId": "cable_act",
                    "providerName": "ACT Cable Network",
                    "price": 0.0,
                    "isActive": True,
                    "createdAt": get_now_iso()
                }

            accounts_list.append({
                "id": item["accountNumber"],
                "accountNumber": item["accountNumber"],
                "customerId": c_id,
                "lcoCustomerId": item["lcoCustomerId"],
                "stbSerialNumber": item["stbSerialNumber"],
                "vcNumber": item["vcNumber"],
                "serviceType": "cable",
                "serviceTypeName": "Cable TV",
                "provider": "act",
                "providerName": "ACT Cable Network",
                "providerId": "cable_act",
                "packageIds": [pkg_slug],
                "packages": [pkg_name],
                "status": item["status"],
                "msoShareDue": item["msoShareDue"],
                "createdAt": item["createdDate"],
                "updatedAt": get_now_iso()
            })

        customers_list = []
        for c in cust_dict.values():
            c["activeServices"] = list(c["activeServices"])
            customers_list.append(c)

        return customers_list, accounts_list, list(packages_map.values())
    else:
        raise FileNotFoundError(f"❌ Customer dataset report not found: {xls_path}")

def run_bulk_upload():
    print("==================================================================")
    print("🚀 SATYA CABLE - FIRESTORE BULK UPLOADER (PYTHON)")
    print("==================================================================")

    db = get_db_client()
    customers_list, accounts_list, packages_list = load_data()

    print(f"\n📊 Summary of Records to Seed:")
    print(f"   - Customers:         {len(customers_list)}")
    print(f"   - Customer Accounts: {len(accounts_list)}")
    print(f"   - Packages:          {len(packages_list)}")
    print(f"   - Service Providers: {len(SERVICE_PROVIDERS)}")

    act_provider = next((p for p in SERVICE_PROVIDERS if p["id"] == "cable_act"), None)
    if act_provider:
        act_provider["packageIds"] = [p["id"] for p in packages_list]

    bulk_writer = db.bulk_writer()
    total_queued = 0

    print("\n🔥 Queueing write operations into Firestore BulkWriter...")

    # 1. Service Providers
    for provider in SERVICE_PROVIDERS:
        ref = db.collection("service_providers").document(provider["id"])
        bulk_writer.set(ref, provider, merge=True)
        total_queued += 1

    # 2. Packages
    for pkg in packages_list:
        root_ref = db.collection("packages").document(pkg["id"])
        bulk_writer.set(root_ref, pkg, merge=True)
        total_queued += 1

        provider_pkg_ref = db.collection("service_providers").document("cable_act").collection("packages").document(pkg["id"])
        bulk_writer.set(provider_pkg_ref, pkg, merge=True)
        total_queued += 1

    # 3. Customers
    for cust in customers_list:
        ref = db.collection("customers").document(cust["id"])
        bulk_writer.set(ref, cust, merge=True)
        total_queued += 1

    # 4. Customer Accounts & Services
    for acct in accounts_list:
        acct_ref = db.collection("customer_accounts").document(acct["id"])
        bulk_writer.set(acct_ref, acct, merge=True)
        total_queued += 1

        svc_ref = db.collection("services").document(acct["id"])
        bulk_writer.set(svc_ref, acct, merge=True)
        total_queued += 1

    print(f"⚡ Flushing {total_queued} batched write operations via BulkWriter...")
    bulk_writer.close()

    print("\n==================================================================")
    print("🎉 FIRESTORE BULK UPLOAD COMPLETE!")
    print(f"   - Total Operations Executed: {total_queued}")
    print("==================================================================")

if __name__ == "__main__":
    run_bulk_upload()

File: scripts/bulk-upload-customers.js
/**
 * Satya Cable & Broadband Network - Bulk Upload Script for Firestore
 *
 * Features:
 * 1. Bulk Operations: Uses Firestore Admin SDK `bulkWriter()` to execute batched
 *    writes efficiently (up to 500 writes per batch automatically queued and flushed),
 *    preventing write operation wastage and hitting rate limits.
 *
 * 2. Service Types & Provider Hierarchy:
 *    - Service Types: 'cable', 'internet', 'internet+iptv', 'apfiber'
 *    - Providers per Service Type:
 *        • cable: 'act' (ACT Cable Network - Current accounts), 'sitv' (SITV Cable)
 *        • internet: 'vbc' (VBC Broadband), 'bsnl' (BSNL Fiber)
 *        • internet+iptv: 'fiber_iptv' (Fiber + IPTV Combo)
 *        • apfiber: 'apsfl' (AP FiberNet)
 *
 * 3. Collection Separation & Bidirectional Relations:
 *    - `customers`: Customer personal details (name, phone, address). Contains `accountIds: string[]`
 *      and `activeServices: string[]` to allow one customer to own multiple service accounts across different types.
 *    - `customer_accounts`: Account-specific connection data (STB serial, VC number, status, provider, serviceType).
 *      Contains `customerId: string` pointing back to `customers/{customerId}`.
 *    - `services`: Mirror/Alias collection of customer_accounts for backward compatibility.
 *
 * 4. Packages & Provider Relations:
 *    - Package details inserted into `service_providers/{providerId}/packages/{packageId}`
 *      and top-level `packages/{packageId}` with relation to provider (`providerId: "cable_act"`).
 *    - `customer_accounts` docs maintain `packageIds: string[]` relation for package queries.
 *
 * Usage:
 *   node scripts/bulk_upload_customers.js
 * Or via npm/pnpm script:
 *   pnpm run seed:customers
 */

const fs = require('node:fs');
const path = require('node:path');
const { initializeApp, cert } = require('firebase-admin/app');
const { getFirestore, FieldValue } = require('firebase-admin/firestore');

// -----------------------------------------------------------------------------
// Configuration & Credentials Setup
// -----------------------------------------------------------------------------
const PROJECT_ID = process.env.FIREBASE_PROJECT_ID || process.env.GCLOUD_PROJECT || 'cableapp-642c4';
const SERVICE_ACCOUNT_FILE = path.join(__dirname, '..', 'cableapp-642c4-firebase-adminsdk-fbsvc-ee5e58c939.json');

let app;
if (process.env.GOOGLE_APPLICATION_CREDENTIALS && fs.existsSync(process.env.GOOGLE_APPLICATION_CREDENTIALS)) {
  app = initializeApp({ credential: cert(process.env.GOOGLE_APPLICATION_CREDENTIALS), projectId: PROJECT_ID });
}
else if (fs.existsSync(SERVICE_ACCOUNT_FILE)) {
  const serviceAccount = JSON.parse(fs.readFileSync(SERVICE_ACCOUNT_FILE, 'utf8'));
  app = initializeApp({ credential: cert(serviceAccount), projectId: PROJECT_ID });
}
else {
  console.log(`ℹ️ Initializing Firebase Admin with default credentials for project: ${PROJECT_ID}`);
  app = initializeApp({ projectId: PROJECT_ID });
}

const db = getFirestore(app);

// -----------------------------------------------------------------------------
// Service Providers Registry Definition
// -----------------------------------------------------------------------------
const SERVICE_PROVIDERS = [
  {
    id: 'cable_act',
    serviceType: 'cable',
    serviceTypeName: 'Cable TV',
    provider: 'act',
    providerName: 'ACT Cable Network',
    isDefault: true,
    packageIds: [],
    updatedAt: new Date().toISOString(),
  },
  {
    id: 'cable_sitv',
    serviceType: 'cable',
    serviceTypeName: 'Cable TV',
    provider: 'sitv',
    providerName: 'SITV Digital Cable',
    isDefault: false,
    packageIds: [],
    updatedAt: new Date().toISOString(),
  },
  {
    id: 'internet_vbc',
    serviceType: 'internet',
    serviceTypeName: 'High-Speed Broadband',
    provider: 'vbc',
    providerName: 'VBC Broadband Network',
    isDefault: false,
    packageIds: [],
    updatedAt: new Date().toISOString(),
  },
  {
    id: 'internet_bsnl',
    serviceType: 'internet',
    serviceTypeName: 'High-Speed Broadband',
    provider: 'bsnl',
    providerName: 'BSNL Bharat Fiber',
    isDefault: false,
    packageIds: [],
    updatedAt: new Date().toISOString(),
  },
  {
    id: 'internet_iptv_combo',
    serviceType: 'internet+iptv',
    serviceTypeName: 'Internet + IPTV Combo',
    provider: 'fiber_iptv',
    providerName: 'Fiber + IPTV Combo Digital',
    isDefault: false,
    packageIds: [],
    updatedAt: new Date().toISOString(),
  },
  {
    id: 'apfiber_apsfl',
    serviceType: 'apfiber',
    serviceTypeName: 'AP FiberNet (APSFL)',
    provider: 'apsfl',
    providerName: 'AP FiberNet Corporation',
    isDefault: false,
    packageIds: [],
    updatedAt: new Date().toISOString(),
  },
];

// Helper to generate a URL/Firestore friendly slug for package names
function getPkgSlug(pkgName) {
  if (!pkgName)
    return 'base_pack';
  return pkgName
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '_')
    .replace(/^_+|_+$/g, '');
}

// -----------------------------------------------------------------------------
// HTML Report Parser (Fallback when JSON files are not present)
// -----------------------------------------------------------------------------
function parseReportHtml(filePath) {
  const htmlContent = fs.readFileSync(filePath, 'utf8');

  // Extract table row matches
  const trMatches = htmlContent.match(/<tr[\s\S]*?<\/tr>/gi) || [];

  const rawAccounts = [];

  for (const tr of trMatches) {
    // Extract td contents
    const tdMatches = tr.match(/<td[\s\S]*?<\/td>/gi);
    if (!tdMatches || tdMatches.length < 10)
      continue;

    // Strip HTML tags and clean up text
    const cleanCells = tdMatches.map((cell) => {
      let text = cell.replace(/<[^>]+>/g, '').trim();
      text = text.replace(/^'/, ''); // remove leading single quote if any
      return text;
    });

    const [sno, acctNo, lcoCustId, stbSerial, vcNo, name, address, createdDate, phone, pkgName, status, msoShareDueStr] = cleanCells;

    if (!acctNo || acctNo.toLowerCase() === 'account no')
      continue;

    rawAccounts.push({
      accountNumber: acctNo,
      lcoCustomerId: lcoCustId || '',
      stbSerialNumber: stbSerial || '',
      vcNumber: vcNo || '',
      name: name || 'Unknown Customer',
      address: address || 'N/A',
      createdDate: createdDate || new Date().toISOString(),
      phone: phone || '919999999999',
      packageName: pkgName || 'Base Pack',
      status: (status || '').toLowerCase().includes('act') ? 'active' : 'inactive',
      msoShareDue: Number.parseFloat(msoShareDueStr || '0') || 0.0,
    });
  }

  return rawAccounts;
}

// -----------------------------------------------------------------------------
// Data Processing & Relationship Mapping
// -----------------------------------------------------------------------------
function processCustomerData() {
  const scriptsDir = __dirname;
  const dataDir = path.join(scriptsDir, 'data');
  const custJsonPath = path.join(dataDir, 'customers.json');
  const svcJsonPath = path.join(dataDir, 'services.json');
  const pkgJsonPath = path.join(dataDir, 'packages.json');
  const xlsPath = path.join(scriptsDir, 'Customer_account_status_report_2026-09-08 14_09_35.xls');

  let rawAccounts = [];

  // Try reading existing JSON or parse raw XLS
  if (fs.existsSync(custJsonPath) && fs.existsSync(svcJsonPath)) {
    console.log(`📦 Loading pre-parsed data from ${dataDir}...`);
    const customersArray = JSON.parse(fs.readFileSync(custJsonPath, 'utf8'));
    const servicesArray = JSON.parse(fs.readFileSync(svcJsonPath, 'utf8'));

    // Return structured collections directly
    return buildCollectionsFromExisting(customersArray, servicesArray);
  }
  else if (fs.existsSync(xlsPath)) {
    console.log(`📦 Parsing customer account status report: ${xlsPath}...`);
    rawAccounts = parseReportHtml(xlsPath);
  }
  else {
    throw new Error(`❌ No data source found! Please make sure ${xlsPath} or JSON files in ${dataDir} exist.`);
  }

  console.log(`   Found ${rawAccounts.length} raw account records.`);

  const DUMMY_PHONES = new Set(['919111111111', '919490712587', '', '919999999999']);
  const dedupKeyToCustId = {};
  let customerCounter = 10001;

  const customersMap = {};
  const accountsList = [];
  const packagesMap = {};

  for (const item of rawAccounts) {
    const phoneClean = (item.phone || '').replace(/\D/g, '');
    const normPhone = DUMMY_PHONES.has(phoneClean) ? '' : phoneClean;
    const normName = (item.name || '').toLowerCase().trim();
    const normAddr = (item.address || '').toLowerCase().trim();

    const dedupKey = normPhone
      ? `phone:${normPhone}`
      : `name_addr:${normName}_${normAddr}`;

    let custId;
    if (!dedupKeyToCustId[dedupKey]) {
      custId = `CUST-${customerCounter++}`;
      dedupKeyToCustId[dedupKey] = custId;

      customersMap[custId] = {
        id: custId,
        customerCode: `SAT-${custId}`,
        name: item.name || 'Unknown Customer',
        phone: item.phone || '919999999999',
        address: item.address || 'N/A',
        accountIds: [], // To be populated with relations
        activeServices: new Set(),
        createdAt: item.createdDate || new Date().toISOString(),
        updatedAt: new Date().toISOString(),
      };
    }
    else {
      custId = dedupKeyToCustId[dedupKey];
    }

    // Link account to customer
    customersMap[custId].accountIds.push(item.accountNumber);
    customersMap[custId].activeServices.add('cable');

    // Process Package
    const pkgName = item.packageName || 'Base Pack';
    const pkgSlug = getPkgSlug(pkgName);
    if (!packagesMap[pkgSlug]) {
      packagesMap[pkgSlug] = {
        id: pkgSlug,
        packageName: pkgName,
        serviceType: 'cable',
        provider: 'act',
        providerId: 'cable_act',
        providerName: 'ACT Cable Network',
        price: 0.0,
        isActive: true,
        createdAt: new Date().toISOString(),
      };
    }

    // Account Document (service: 'cable', provider: 'act')
    const accountDoc = {
      id: item.accountNumber,
      accountNumber: item.accountNumber,
      customerId: custId, // Relation back to Customer
      lcoCustomerId: item.lcoCustomerId || '',
      stbSerialNumber: item.stbSerialNumber || '',
      vcNumber: item.vcNumber || '',
      serviceType: 'cable',
      serviceTypeName: 'Cable TV',
      provider: 'act',
      providerName: 'ACT Cable Network',
      providerId: 'cable_act',
      packageIds: [pkgSlug],
      packages: [pkgName],
      status: item.status || 'active',
      msoShareDue: item.msoShareDue || 0.0,
      createdAt: item.createdDate || new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };

    accountsList.push(accountDoc);
  }

  // Convert activeServices set to array for Firestore
  const customersList = Object.values(customersMap).map(c => ({
    ...c,
    activeServices: Array.from(c.activeServices),
  }));

  const packagesList = Object.values(packagesMap);

  return { customersList, accountsList, packagesList };
}

function buildCollectionsFromExisting(customersArray, servicesArray) {
  const packagesMap = {};
  const customerAccountIdsMap = {};
  const customerServicesMap = {};

  const accountsList = servicesArray.map((s) => {
    const pkgName = (s.packages && s.packages[0]) ? s.packages[0] : 'Base Pack';
    const pkgSlug = (s.packageIds && s.packageIds[0]) ? s.packageIds[0] : getPkgSlug(pkgName);

    if (!packagesMap[pkgSlug]) {
      packagesMap[pkgSlug] = {
        id: pkgSlug,
        packageName: pkgName,
        serviceType: s.serviceCategory || 'cable',
        provider: s.provider || 'act',
        providerId: s.providerId || 'cable_act',
        providerName: s.providerName || 'ACT Cable Network',
        price: 0.0,
        isActive: true,
        createdAt: new Date().toISOString(),
      };
    }

    const custId = s.customerId;
    if (!customerAccountIdsMap[custId]) {
      customerAccountIdsMap[custId] = [];
      customerServicesMap[custId] = new Set();
    }
    customerAccountIdsMap[custId].push(s.accountNumber || s.id);
    customerServicesMap[custId].add(s.serviceCategory || 'cable');

    return {
      id: s.id || s.accountNumber,
      accountNumber: s.accountNumber || s.id,
      customerId: custId,
      lcoCustomerId: s.lcoCustomerId || '',
      stbSerialNumber: s.stbSerialNumber || '',
      vcNumber: s.vcNumber || '',
      serviceType: s.serviceCategory || 'cable',
      serviceTypeName: s.serviceCategoryName || 'Cable TV',
      provider: s.provider || 'act',
      providerId: s.providerId || 'cable_act',
      providerName: s.providerName || 'ACT Cable Network',
      packageIds: [pkgSlug],
      packages: [pkgName],
      status: s.status || 'active',
      msoShareDue: s.msoShareDue || 0.0,
      createdAt: s.createdAt || new Date().toISOString(),
      updatedAt: new Date().toISOString(),
    };
  });

  const customersList = customersArray.map(c => ({
    ...c,
    accountIds: customerAccountIdsMap[c.id] || [],
    activeServices: Array.from(customerServicesMap[c.id] || ['cable']),
    updatedAt: new Date().toISOString(),
  }));

  const packagesList = Object.values(packagesMap);

  return { customersList, accountsList, packagesList };
}

// -----------------------------------------------------------------------------
// Bulk Upload Execution via BulkWriter
// -----------------------------------------------------------------------------
async function runBulkUpload() {
  console.log('==================================================================');
  console.log('🚀 SATYA CABLE - FIRESTORE BULKWRITER CUSTOMER DATA UPLOADER');
  console.log('==================================================================');

  const { customersList, accountsList, packagesList } = processCustomerData();

  console.log(`\n📊 Data Overview to Upload:`);
  console.log(`   - Customers Collection:        ${customersList.length} documents`);
  console.log(`   - Accounts & Services:         ${accountsList.length} documents`);
  console.log(`   - Packages Collection:         ${packagesList.length} documents`);
  console.log(`   - Service Providers Registry:  ${SERVICE_PROVIDERS.length} documents`);

  // Initialize BulkWriter
  const bulkWriter = db.bulkWriter();
  let totalQueued = 0;
  let successCount = 0;
  let errorCount = 0;

  bulkWriter.onWriteResult((ref) => {
    successCount++;
    if (successCount % 250 === 0 || successCount === totalQueued) {
      console.log(`   ✅ Firestore BulkWriter Progress: ${successCount}/${totalQueued} docs committed...`);
    }
  });

  bulkWriter.onWriteError((ref, error, attempts) => {
    errorCount++;
    console.error(`   ❌ Write error for ${ref.path} (Attempt ${attempts}):`, error.message);
    if (error.message.includes('429') || error.message.includes('RESOURCE_EXHAUSTED')) {
      console.warn(`   ⚠️ Rate limit / quota reached. BulkWriter will auto-retry up to limit.`);
    }
    return attempts <= 3; // Retry up to 3 times for transient errors
  });

  console.log(`\n🔥 Queueing documents for BulkWriter...`);

  // 1. Upload Service Providers Registry
  const actProvider = SERVICE_PROVIDERS.find(p => p.id === 'cable_act');
  if (actProvider) {
    actProvider.packageIds = packagesList.map(p => p.id);
  }

  for (const provider of SERVICE_PROVIDERS) {
    const ref = db.collection('service_providers').doc(provider.id);
    bulkWriter.set(ref, provider, { merge: true });
    totalQueued++;
  }

  // 2. Upload Packages (Top-level & Provider Subcollection)
  for (const pkg of packagesList) {
    // Root packages collection
    const rootRef = db.collection('packages').doc(pkg.id);
    bulkWriter.set(rootRef, pkg, { merge: true });
    totalQueued++;

    // Provider subcollection packages (service_providers/cable_act/packages/{pkgId})
    const providerPkgRef = db
      .collection('service_providers')
      .doc('cable_act')
      .collection('packages')
      .doc(pkg.id);
    bulkWriter.set(providerPkgRef, pkg, { merge: true });
    totalQueued++;
  }

  // 3. Upload Customers Collection
  for (const cust of customersList) {
    const ref = db.collection('customers').doc(cust.id);
    bulkWriter.set(ref, cust, { merge: true });
    totalQueued++;
  }

  // 4. Upload Customer Accounts Collection (`customer_accounts` & mirror to `services`)
  for (const acct of accountsList) {
    // Primary customer_accounts collection
    const acctRef = db.collection('customer_accounts').doc(acct.id);
    bulkWriter.set(acctRef, acct, { merge: true });
    totalQueued++;

    // Mirror to services collection for legacy query compatibility
    const svcRef = db.collection('services').doc(acct.id);
    bulkWriter.set(svcRef, acct, { merge: true });
    totalQueued++;
  }

  console.log(`⚡ Flushing ${totalQueued} queued write operations to Firestore via BulkWriter...`);

  await bulkWriter.close();

  console.log('\n==================================================================');
  console.log('🎉 BULK UPLOAD TO FIRESTORE COMPLETE!');
  console.log(`   - Total Operations Committed: ${successCount} / ${totalQueued}`);
  console.log(`   - Customers Created/Updated:  ${customersList.length}`);
  console.log(`   - Accounts Created/Updated:   ${accountsList.length}`);
  console.log(`   - Packages Created/Updated:   ${packagesList.length}`);
  console.log(`   - Service Providers Seeded:   ${SERVICE_PROVIDERS.length}`);
  console.log('==================================================================');
}

// Execute main upload function
runBulkUpload().catch((err) => {
  console.error('\n⛔ FATAL ERROR DURING BULK UPLOAD:', err);
  process.exit(1);
});
